### PR TITLE
Add Hardware Accelerated Checksums

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

Added support for hardware accelerated Crc32 and Adler32 checksum generation both used by our png codecs. They're fast!

Benchmarks.

**Adler32**

|               Method |       Runtime | Count |        Mean |       Error |    StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |-------------- |------ |------------:|------------:|----------:|------:|--------:|------:|------:|------:|----------:|
| SharpZipLibCalculate |    .NET 4.7.2 |  1024 |   793.18 ns |   775.66 ns | 42.516 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate |    .NET 4.7.2 |  1024 |   384.86 ns |    15.64 ns |  0.857 ns |  0.49 |    0.03 |     - |     - |     - |         - |
|                      |               |       |             |             |           |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 2.1 |  1024 |   790.31 ns |   353.34 ns | 19.368 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 2.1 |  1024 |   465.28 ns |   652.41 ns | 35.761 ns |  0.59 |    0.03 |     - |     - |     - |         - |
|                      |               |       |             |             |           |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 3.1 |  1024 |   877.25 ns |    97.89 ns |  5.365 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 3.1 |  1024 |    45.60 ns |    13.28 ns |  0.728 ns |  0.05 |    0.00 |     - |     - |     - |         - |
|                      |               |       |             |             |           |       |         |       |       |       |           |
| SharpZipLibCalculate |    .NET 4.7.2 |  2048 | 1,537.04 ns |   428.44 ns | 23.484 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate |    .NET 4.7.2 |  2048 |   849.76 ns | 1,066.34 ns | 58.450 ns |  0.55 |    0.04 |     - |     - |     - |         - |
|                      |               |       |             |             |           |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 2.1 |  2048 | 1,616.97 ns |   276.70 ns | 15.167 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 2.1 |  2048 |   790.77 ns |   691.71 ns | 37.915 ns |  0.49 |    0.03 |     - |     - |     - |         - |
|                      |               |       |             |             |           |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 3.1 |  2048 | 1,735.11 ns | 1,374.22 ns | 75.325 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 3.1 |  2048 |    87.80 ns |    56.84 ns |  3.116 ns |  0.05 |    0.00 |     - |     - |     - |         - |
|                      |               |       |             |             |           |       |         |       |       |       |           |
| SharpZipLibCalculate |    .NET 4.7.2 |  4096 | 3,054.53 ns |   796.41 ns | 43.654 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate |    .NET 4.7.2 |  4096 | 1,538.90 ns |   487.02 ns | 26.695 ns |  0.50 |    0.01 |     - |     - |     - |         - |
|                      |               |       |             |             |           |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 2.1 |  4096 | 3,223.48 ns |    32.32 ns |  1.771 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 2.1 |  4096 | 1,547.60 ns |   309.72 ns | 16.977 ns |  0.48 |    0.01 |     - |     - |     - |         - |
|                      |               |       |             |             |           |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 3.1 |  4096 | 3,672.33 ns | 1,095.81 ns | 60.065 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 3.1 |  4096 |   159.44 ns |    36.31 ns |  1.990 ns |  0.04 |    0.00 |     - |     - |     - |         - |

**Crc32**

|               Method |       Runtime | Count |         Mean |        Error |       StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |-------------- |------ |-------------:|-------------:|-------------:|------:|--------:|------:|------:|------:|----------:|
| SharpZipLibCalculate |    .NET 4.7.2 |  1024 |  3,067.24 ns |    769.25 ns |    42.165 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate |    .NET 4.7.2 |  1024 |  2,546.86 ns |  1,106.36 ns |    60.643 ns |  0.83 |    0.02 |     - |     - |     - |         - |
|                      |               |       |              |              |              |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 2.1 |  1024 |  3,377.15 ns |  3,903.41 ns |   213.959 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 2.1 |  1024 |  2,524.25 ns |  2,220.97 ns |   121.739 ns |  0.75 |    0.04 |     - |     - |     - |         - |
|                      |               |       |              |              |              |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 3.1 |  1024 |  3,980.60 ns |  8,497.37 ns |   465.769 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 3.1 |  1024 |     78.68 ns |     69.82 ns |     3.827 ns |  0.02 |    0.00 |     - |     - |     - |         - |
|                      |               |       |              |              |              |       |         |       |       |       |           |
| SharpZipLibCalculate |    .NET 4.7.2 |  2048 |  7,934.29 ns | 42,550.13 ns | 2,332.316 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate |    .NET 4.7.2 |  2048 |  5,437.81 ns | 12,760.51 ns |   699.447 ns |  0.71 |    0.10 |     - |     - |     - |         - |
|                      |               |       |              |              |              |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 2.1 |  2048 |  6,008.05 ns |    621.37 ns |    34.059 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 2.1 |  2048 |  4,791.50 ns |  3,894.94 ns |   213.495 ns |  0.80 |    0.04 |     - |     - |     - |         - |
|                      |               |       |              |              |              |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 3.1 |  2048 |  5,900.06 ns |  1,344.70 ns |    73.707 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 3.1 |  2048 |    103.12 ns |     15.66 ns |     0.859 ns |  0.02 |    0.00 |     - |     - |     - |         - |
|                      |               |       |              |              |              |       |         |       |       |       |           |
| SharpZipLibCalculate |    .NET 4.7.2 |  4096 | 12,422.59 ns |  1,308.01 ns |    71.696 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate |    .NET 4.7.2 |  4096 | 10,524.63 ns |  6,267.56 ns |   343.546 ns |  0.85 |    0.03 |     - |     - |     - |         - |
|                      |               |       |              |              |              |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 2.1 |  4096 | 11,888.00 ns |  1,059.25 ns |    58.061 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 2.1 |  4096 |  9,806.24 ns |    241.91 ns |    13.260 ns |  0.82 |    0.00 |     - |     - |     - |         - |
|                      |               |       |              |              |              |       |         |       |       |       |           |
| SharpZipLibCalculate | .NET Core 3.1 |  4096 | 12,181.28 ns |  1,974.68 ns |   108.239 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|   SixLaborsCalculate | .NET Core 3.1 |  4096 |    192.39 ns |     10.27 ns |     0.563 ns |  0.02 |    0.00 |     - |     - |     - |         - |

### CODE REVIEW

File: src/ImageSharp/Formats/Png/Zlib/Adler32.cs [hunk 1]
```diff
@@ -4,145 +4,283 @@
...
+                k = length < NMAX ? length : NMAX;
+                length -= k;
+
+                while (k >= 16)
                 {
-                    n = count;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
```
Comment (saucecontrol) [inline src/ImageSharp/Formats/Png/Zlib/Adler32.cs:237]: JIT doesn't do great with code like this.

On legacy jit32, it emits this: 

```asm
    L0031: movzx eax, byte [ebx+eax]
    L0035: add edx, eax
    L0037: add ecx, edx
    L0039: mov eax, esi
    L003b: inc esi
    L003c: movzx eax, byte [ebx+eax]
    L0040: add edx, eax
    L0042: add ecx, edx
    L0044: mov eax, esi
    L0046: inc esi
    …
```

RyuJIT does a bit better by not *actually* incrementing `index` each time, but it emits an extra `mov` for each block because it's confused by your two variables:

```asm
    L0031: movzx r9d, byte ptr [rax+1]
    L0036: add r9d, ecx
    L0039: mov ecx, r9d
    L003c: add r8d, ecx
    L003f: movzx r9d, byte ptr [rax+2]
    L0044: add r9d, ecx
    L0047: mov ecx, r9d
    L004a: add r8d, ecx
    …
```

Writing it more like the C code:

```C#
    byte* pbuff = pbuffer + index;
    if (length >= 16) 
    { 
        s2 += (s1 += pbuff[0]); 
        s2 += (s1 += pbuff[1]); 
        s2 += (s1 += pbuff[2]); 
        ...
    }
```

gives better codegen in both:

```asm
    L0042: movzx r9d, byte ptr [rcx+1]
    L0047: add eax, r9d
    L004a: add r8d, eax
    L004d: movzx r9d, byte ptr [rcx+2]
    L0052: add eax, r9d
    L0055: add r8d, eax
```

[SharpLab for both variants](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9Jq049+ggJI8ZEMfKhKV6mpoDMTUgyEMA3jQYem9jrx4MAsgAUsmLYXAA8wACeGDAAfAzAHABmyTIAlO6ebtSeeQywyYkxMIkpaVBsMEUAvAXVATD40FH+2FC4ABbYkiwA4jAYVRUwXCqBSakZagw2ufkevhgM0lwA5hidDHWTFSwAMqMbnRrzC1kLAPSXDJuqpQDuUpIM3Qqlze8AJgzJUKK3TqlWQggCi9kkEFMszO+R8flw5G2DHIpwWHnhy1wjjqqIu+SWDF8XxgCGRdDR6PxeV4RUCq2ODDiuKQ6Rh6Nc7I5nkRDAA1HUAKpcXDYNIsACCXy+gUKZSmlWqaCJXBJCD5fPSM2p6Ox/LqiO1sI5vIFDGFovFUplct2MmGyuJpI1Wq53IYerNhrd3NNQpFYsE1tlDTtiuSjtVzs1RvdHk9BtRPpNSLNFsDkulIaKYYdKrVLtjcYTHqTOoWfvNAatWdt5XtSvz0dd5fyJe9rbylfTNZtofr4cjBZjyd1ji9ZeNutT/stQdr/YVead6pHnZ548TRfd3er877OYHy6jq5bU4rm9L299M6rc8zB/lFWPw7PcY9l4757bt57++zT4NhGTanteJqfpO767vewZ1kujYroWo4XvqV7IT+qF/g+AG5ghJ5Ieu8YQWB06YXu2Fwc+eGviRKETrRGFpuRsGLlRwGIWu35dsR6Fdr+zELoe8HsfhnFQTxhGlmRMGCYBg4gQRXEbqhX7vgymwMHALIMQAvjCrYPJ0UilPSRybHAcBMgwdBsq2Li8cpE7SRmLFCWxQ7Nrp6GtsQADsUmEAwgR6uE4QoqylIeDpcx5MQ3h7gwmIMAAcsEoQRNEsQJGGmSwjkHLJLwCAwD8EwlAAVAwYhhsiOWuDF6KEupWw7AOBxmScrattcgL3AwTySC8bwfBA3y/P8+C9QwIKyOCKxQmI6FJbyuKRXChJ6qtraEiu5JrRyraZTAlXVeUyKnQq/Igftni0kFzVMiytnfvZkklsFt4XckADadAALpvsWl4fahX3feQAMMdxqEg2aYOkJDDlETDlZg7YiNvcDqO7N9KAY0pyNmrDdRgwArPj4ko59ONIBTQNU6DOO+XTO5Y9T5TfQAHCzvps4zHMAJw8+BDNwzj5D/YDrOiyT4sQ1LvMy1V4sIwrItE9jHPkOjatjkrYPkHjut0XUxPK1r5PG/kX2oeQSA3QsD1aeFXkNQshnGfdHUWVZNnoa9BMfkr5VfYpHJ6V135+QFQUhWFdtaq20WwsnOlAA===)

It's also not clear that unroll by 16 is ideal for modern processors.  This may be an outdated optimization from zlib.  Might be worth testing unroll by 8 or even 4 to see how they do if you haven't already.

File: src/ImageSharp/Formats/Png/Zlib/Crc32.cs [hunk 2]
```diff
@@ -4,151 +4,206 @@
...
-        public void Update(int value)
-        {
-            this.crc ^= CrcSeed;
-            this.crc = CrcTable[(this.crc ^ value) & 0xFF] ^ (this.crc >> 8);
-            this.crc ^= CrcSeed;
+            fixed (byte* bufferPtr = &buffer[0])
+            fixed (ulong* k1k2Ptr = &k1k2[0])
+            fixed (ulong* k3k4Ptr = &k3k4[0])
+            fixed (ulong* k5k0Ptr = &k5k0[0])
+            fixed (ulong* polyPtr = &poly[0])
```
Comment (saucecontrol) [inline src/ImageSharp/Formats/Png/Zlib/Crc32.cs:86]: This is another place where the ROS<byte> trick would be of benefit. Instead of 4 separate unmanaged pointers, you could have a single pointer to a block containing all 4 vector values.

File: tests/ImageSharp.Tests/Formats/Png/PngDecoderTests.Chunks.cs [hunk 3]
```diff
@@ -80,17 +84,54 @@ public void Decode_IncorrectCRCForCriticalChunk_ExceptionIsThrown(uint chunkType
...
-        public void CalculateCrc_Works()
+        public void CalculateCrc_Works_LongerRun()
         {
-            // arrange
-            var data = new byte[] { 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07 };
-            var crc = new Crc32();
-
-            // act
-            crc.Update(data);
+            // Longer run, enough to require moving the point in SIMD implementation with
```
Comment (antonfirsov) [inline tests/ImageSharp.Tests/Formats/Png/PngDecoderTests.Chunks.cs:89]: Are there any further corner cases worth of testing?

Comment (JimBobSquarePants) [reply tests/ImageSharp.Tests/Formats/Png/PngDecoderTests.Chunks.cs]: Yup, The current test isn't great, I'll add better reference tests

File: src/ImageSharp/Formats/Png/Zlib/Crc32.cs (hunk 2 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Png/Zlib/Crc32.cs]: I've updated the code to use a static readonly array as I wasn't sure the ROS optimizations apply for anything other than byte?

**Comment by JimBobSquarePants** [comment]
@saucecontrol Thanks for the review! Some great improvement ideas there. 👍

File: src/ImageSharp/Formats/Png/Zlib/Crc32.cs (hunk 2 above)
Comment (saucecontrol) [reply src/ImageSharp/Formats/Png/Zlib/Crc32.cs]: Yep, the compiler won't optimize anything bigger than byte because of endianness concerns. Since this code is processor-specific, you can outsmart the compiler by breaking the ulongs up into bytes (and reversing their byte order, ofc)

Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Png/Zlib/Crc32.cs]: I'll leave that for now since the massive hashing improvements have actually made less of an impact than I'd hoped. I'm going to profile now and find other soft targets.
